### PR TITLE
fix a legend bug for scatter faceted plot

### PR DESCRIPTION
**Pull request by moontrip**
While thinking and checking of a potential scatter plot R-square table with a couple of example cases, I found that there was a bug for some specific data for the case of faceted plots. 

Just made a new ticket so this PR will address #741 

For an example like below with the plot options of Best fit with raw (or Smoothed mean with raw) using GEMS1 Study control, the custom legend showed 5 Raw and 3 Best fit. It is because the condition to find the legend list is based on the first non-undefined data. However, when I checked whole faceted data, it is found that actually all 5 Best fits exist throughout countries, thus it is necessary to find whole 5 Best fits by going through whole data, like below screenshot. While fixing the issue, it was inevitable to use complex conditions to properly order the legend items. Perhaps that part may be revisited for better logic/scheme in the future, especially considering the B55 release date, but at least this PR would work.

# before fix
scatter-facet-bug

# after fix
scatter-bug-fix

### CODE REVIEW

**Comment by moontrip** [comment]
Since the freezing will happen today, I listed @dmfalke as well in case @bobular is not available. Certainly @jtlong3rd is also a reviewer too 😄

**Review by bobular — APPROVED** [review]
Hi DK - great catch and good solution.  Though I will admit I didn't spend the time to fully understand the new sorting logic.

What I did instead, which may reduce technical debt in the future, is get rid of the `as any[]` and used a trick that Dave showed me (adding the type guard to the `!= null` filter).  Given that `uniqBy` only cares about the `name` prop, I think it's fine to flatten all the facet series "whole" pulling out just the `{ name, mode, fill }` props. That avoids the TS issues.

**Comment by moontrip** [comment]
@bobular yes I just saw your commits. I was also thinking of the case to use full data, but actually for legend, the three object keys are required so I just remade object containing the three keys/values only to save some memory. But it may not be that optimal, and your change looks great! 👍 

Thank you for your reviews and comments!

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 1]
```diff
@@ -423,23 +424,40 @@ function ScatterplotViz(props: VisualizationProps) {
...
+
     const legendData = !isFaceted(allData)
       ? allData?.series
-      : allData.facets.find(
-          ({ data }) => data != null && data.series.length > 0
-        )?.data?.series;
+      : uniqBy(
+          allData.facets
+            .map((facet) => facet?.data?.series)
+            .flat()
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:442]:
```suggestion
            .flatMap((facet) => facet?.data?.series)
```

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 2]
```diff
@@ -449,9 +467,118 @@ function ScatterplotViz(props: VisualizationProps) {
...
+    let count = 0;
+
+    // new array of objects based on legendData array (a kind of partial sort)
+    // raw data is in the correct order, but not always for smoothed mean, CI, and Best fit
+    // e.g., showing 'data1' (raw), 'data2' (raw), 'data2, Best fit', 'data1, Best fit' in order
+    // needed to have complex conditions for treating Smoothed mean, CI, Best fit, No data, etc.
+    const sortedLegendData =
+      isFaceted(allData) && vizConfig.valueSpecConfig !== 'Raw'
+        ? legendData
+            ?.map((dataItem, index) => {
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:485]:
```suggestion
            ?.flatMap((dataItem, index) => {
```

File: src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx [hunk 3]
```diff
@@ -449,9 +467,118 @@ function ScatterplotViz(props: VisualizationProps) {
...
+                        return element.name?.includes('No data' + CI95SUFFIX);
+                      });
+                    }
+                  }
+                }
+              }
+              return [];
+              // observed No data is often undefined during data loading by toggling showMissingness for large scatter data/graphics
+            })
+            .flat()
```
Comment (jtlong3rd) [inline src/lib/core/components/visualizations/implementations/ScatterplotVisualization.tsx:575]:
```suggestion
```

**Review by jtlong3rd — APPROVED** [review]
Just sneaking in the usual small suggestions before we merge. 😄

**Comment by moontrip** [comment]
@jtlong3rd Thank you for your reviews! 👍 Yes, map( ) + flat() = flatmap() 😄 Applied your changes

@dmfalke since Bob and Jamie reviewed this PR, I will merge this.

**Pull request merged**